File: download/mastery-engine/sdks/java/src/main/java/com/masteryos/MasteryOS.java
package com.masteryos;

/**
 * MasteryOS Java SDK — official client library.
 *
 * Usage:
 *   MasteryOS client = new MasteryOS.Builder().apiKey("your-key").build();
 *   Map<String, Object> dashboard = client.learning().getDashboard();
 */

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Map;

public class MasteryOS {
    private static final String VERSION = "1.0.0";
    private static final String DEFAULT_BASE_URL = "https://api.masteryos.com";

    private final String apiKey;
    private final String baseUrl;
    private final HttpClient httpClient;
    private final int maxRetries;

    private final Learning learning;
    private final Auth auth;

    private MasteryOS(Builder builder) {
        this.apiKey = builder.apiKey;
        this.baseUrl = builder.baseUrl != null ? builder.baseUrl : DEFAULT_BASE_URL;
        this.maxRetries = builder.maxRetries != null ? builder.maxRetries : 3;
        this.httpClient = HttpClient.newBuilder()
                .timeout(Duration.ofSeconds(builder.timeout != null ? builder.timeout : 30))
                .build();
        this.learning = new Learning(this);
        this.auth = new Auth(this);
    }

    public Learning learning() { return learning; }
    public Auth auth() { return auth; }

    public static class Builder {
        private String apiKey;
        private String baseUrl;
        private Integer timeout;
        private Integer maxRetries;

        public Builder apiKey(String apiKey) { this.apiKey = apiKey; return this; }
        public Builder baseUrl(String baseUrl) { this.baseUrl = baseUrl; return this; }
        public Builder timeout(int seconds) { this.timeout = seconds; return this; }
        public Builder maxRetries(int retries) { this.maxRetries = retries; return this; }

        public MasteryOS build() {
            if (apiKey == null || apiKey.isEmpty()) {
                throw new IllegalArgumentException("API key is required");
            }
            return new MasteryOS(this);
        }
    }

    String request(String method, String path, String body) throws Exception {
        String url = baseUrl + path;
        HttpRequest.Builder reqBuilder = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + apiKey)
                .header("Content-Type", "application/json")
                .header("User-Agent", "masteryos-java/" + VERSION);

        if (body != null) {
            reqBuilder.method(method, HttpRequest.BodyPublishers.ofString(body));
        } else {
            reqBuilder.method(method, HttpRequest.BodyPublishers.noBody());
        }

        Exception lastError = null;
        for (int attempt = 0; attempt <= maxRetries; attempt++) {
            try {
                HttpResponse<String> response = httpClient.send(reqBuilder.build(), HttpResponse.BodyHandlers.ofString());
                if (response.statusCode() >= 500 && attempt < maxRetries) {
                    Thread.sleep((long) Math.pow(2, attempt) * 1000);
                    continue;
                }
                if (response.statusCode() >= 400) {
                    throw new RuntimeException("API error [" + response.statusCode() + "]: " + response.body());
                }
                return response.body();
            } catch (Exception e) {
                lastError = e;
                if (attempt < maxRetries) {
                    Thread.sleep((long) Math.pow(2, attempt) * 1000);
                }
            }
        }
        throw new RuntimeException("Request failed after retries", lastError);
    }

    public static class Learning {
        private final MasteryOS client;
        Learning(MasteryOS client) { this.client = client; }

        public String getDashboard() throws Exception {
            return client.request("GET", "/api/v1/learning/dashboard", null);
        }

        public String startSession(String subjectId) throws Exception {
            return client.request("POST", "/api/v1/learning/sessions",
                "{\"subject_id\":\"" + subjectId + "\",\"intent\":\"mixed\",\"target_question_count\":10}");
        }
    }

    public static class Auth {
        private final MasteryOS client;
        Auth(MasteryOS client) { this.client = client; }

        public String login(String email, String password) throws Exception {
            return client.request("POST", "/api/v1/auth/login",
                "{\"email\":\"" + email + "\",\"password\":\"" + password + "\"}");
        }
    }
}

File: download/mastery-engine/sdks/go/masteryos.go
// Package masteryos provides a Go client for the MasteryOS API.
//
// Usage:
//   client := masteryos.New("your-api-key")
//   dashboard, err := client.Learning.GetDashboard()
package masteryos

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

const Version = "1.0.0"
const DefaultBaseURL = "https://api.masteryos.com"

// Client is the main MasteryOS API client.
type Client struct {
	APIKey     string
	BaseURL    string
	HTTPClient *http.Client
	MaxRetries int

	Learning *LearningService
	Auth     *AuthService
}

// New creates a new MasteryOS client with the given API key.
func New(apiKey string) *Client {
	c := &Client{
		APIKey:     apiKey,
		BaseURL:    DefaultBaseURL,
		HTTPClient: &http.Client{Timeout: 30 * time.Second},
		MaxRetries: 3,
	}
	c.Learning = &LearningService{client: c}
	c.Auth = &AuthService{client: c}
	return c
}

// APIError represents an API error response.
type APIError struct {
	StatusCode int
	Message    string
	Code       string
}

func (e *APIError) Error() string {
	return fmt.Sprintf("[%d] %s", e.StatusCode, e.Message)
}

func (c *Client) do(method, path string, body interface{}) (map[string]interface{}, error) {
	var reqBody io.Reader
	if body != nil {
		jsonBody, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		reqBody = bytes.NewReader(jsonBody)
	}

	url := c.BaseURL + path
	var lastErr error

	for attempt := 0; attempt <= c.MaxRetries; attempt++ {
		req, err := http.NewRequest(method, url, reqBody)
		if err != nil {
			return nil, err
		}
		req.Header.Set("Authorization", "Bearer "+c.APIKey)
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("User-Agent", "masteryos-go/"+Version)

		resp, err := c.HTTPClient.Do(req)
		if err != nil {
			lastErr = err
			if attempt < c.MaxRetries {
				time.Sleep(time.Duration(1<<attempt) * time.Second)
				continue
			}
			return nil, err
		}
		defer resp.Body.Close()

		var result map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&result)

		if resp.StatusCode >= 400 {
			return nil, &APIError{StatusCode: resp.StatusCode, Message: fmt.Sprintf("%v", result["detail"])}
		}
		return result, nil
	}
	return nil, lastErr
}

// LearningService provides learning endpoints.
type LearningService struct{ client *Client }

func (s *LearningService) GetDashboard() (map[string]interface{}, error) {
	return s.client.do("GET", "/api/v1/learning/dashboard", nil)
}

func (s *LearningService) StartSession(subjectID string) (map[string]interface{}, error) {
	return s.client.do("POST", "/api/v1/learning/sessions", map[string]interface{}{
		"subject_id": subjectID, "intent": "mixed", "target_question_count": 10,
	})
}

// AuthService provides authentication endpoints.
type AuthService struct{ client *Client }

func (s *AuthService) Login(email, password string) (map[string]interface{}, error) {
	return s.client.do("POST", "/api/v1/auth/login", map[string]interface{}{
		"email": email, "password": password,
	})
}
